### PR TITLE
Add `records.list` rpc function

### PR DESCRIPTION
**Pull request by mathemancer**
Related to #3690 

Adds an initial `records.list` RPC function.

**Technical details**

The `filter`, `group`, and `search` kwargs aren't wired up yet, but can be submitted.

As usual, the included documentation should be sufficient to describe the functions for review.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by mathemancer** [comment]
> Looks good!
> 
> * I skimmed the code. No red flags stuck out to me. Though I imagine Anish will go over in it more detail.
> * I ran it and verified that it works.
> * I stress-tested the `limit`, `offset`, and `order` parameters a bit and couldn't find anything amiss.
> * The general structure of the output looks good.
> 
> The only (slight) concern I have is this...
> 
> The cell values for some types are not formatted the same way as in the REST API. For example:
> 
> REST	RPC
> `1960-07-24 AD`	`1960-07-24`
> `2014-04-21T17:30:44.0 AD`	`2014-04-21T17:30:44`
> `P0Y0M0DT777H55M39S`	`777:55:39`
> It's hard for me to quickly identify whether the change in formatting will be problematic or not. We would certainly notice this later down the road during QA though, so I'm okay kicking this can down the road and seeing what problems pop up. They'll be easier to address when we have this new RPC API better integrated into the front end, because at that point we'll be able to make a more informed decision about the quickest path to fixing any issues (i.e. front end or back end).

I'll look into the formatting. Great catch, thank you!

**Comment by mathemancer** [comment]
@Anish9901 When you review: I just ended up emulating the current REST endpoint responses w.r.t. the formatting. This means:

- No formatting of types within arrays, and
- The `interval` type formatting doesn't really work in the front end (same as before).

I think we can address these issues as feature requests in a future PR.

File: db/sql/00_msar.sql [hunk 1]
```diff
@@ -3228,3 +3228,222 @@ BEGIN
...
+
+column_name AS "2", another_column_name AS "4"
+
+Args:
+  tab_id: The OID of the table containing the columns to select.
+*/
+SELECT string_agg(format('msar.format_data(%I) AS %I', attname, attnum), ', ')
+FROM pg_catalog.pg_attribute
+WHERE
+  attrelid = tab_id AND attnum > 0 AND has_column_privilege(attrelid, attnum, 'SELECT');
```
Comment (Anish9901) [inline db/sql/00_msar.sql:3396]: This breaks for a table from which previously existing column(s) have been dropped.
Here is the specific error:
```py
mathesar=# select msar.list_records_from_table('"Library Management"."Items"'::regclass::oid, NULL, NULL, NULL, NULL, NULL, NULL);
ERROR:  column "........pg.dropped.4........" does not exist
LINE 5: ...data("Acquisition Date") AS "3", msar.format_data("........p...
                                                             ^
QUERY:  
    WITH count_cte AS (
      SELECT count(1) AS count FROM "Library Management"."Items"
    ), results_cte AS (
      SELECT msar.format_data(id) AS "1", msar.format_data("Barcode") AS "2", msar.format_data("Acquisition Date") AS "3", msar.format_data("........pg.dropped.4........") AS "4", msar.format_data("Book") AS "5" FROM "Library Management"."Items" ORDER BY "1" ASC LIMIT NULL OFFSET NULL
    )
    SELECT jsonb_build_object(
      'results', jsonb_agg(row_to_json(results_cte.*)),
      'count', max(count_cte.count)
    )
    FROM results_cte, count_cte
    
CONTEXT:  PL/pgSQL function msar.list_records_from_table(oid,integer,integer,jsonb,jsonb,jsonb,jsonb) line 19 at EXECUTE
```

**Review by Anish9901 — CHANGES_REQUESTED** [review]
The PR looks great overall @mathemancer, and I find the idea of using the attnums as column identifiers instead of referring to them by their names brilliant! However, I think I have found a pretty significant bug here which is worth fixing.

File: db/sql/00_msar.sql (hunk 1 above)
Comment (mathemancer) [reply db/sql/00_msar.sql]: Nice catch! I've pushed a fix and a test. Should be fine now.

**Review by Anish9901 — APPROVED** [review]

**Pull request merged**